A practical introduction to Pandas
===================

We have been asked to *analyze an otp dataset*, without much more information. This kind of scenario is more common than you might imagine!

## 1) Quickly examine the files in ~/Data/us_dot/otp. What do they contain, in both technical and functional terms? (Use any tool you want)

In [57]:
path_to_zips = '/home/dsc/Data/us_dot/otp/'

For this example, we are going to begin using the shell from within this notebook, with the ! notation:

A ! sign before a line tells the notebook to send that line straight away to the underlying OS. 

\* Note that we can substitute python variables into the shell command. We do that by surrounding the name of the variable with curly braces ({}). That's what we are going to do with the `path_to_files` variable

### Check what the files contain

In [58]:
!unzip -l {path_to_zips}On_Time_On_Time_Performance_2015_1.zip

Archive:  /home/dsc/Data/us_dot/otp/On_Time_On_Time_Performance_2015_1.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
211633432  04-16-2015 20:02   On_Time_On_Time_Performance_2015_1.csv
    12054  04-16-2015 20:02   readme.html
---------                     -------
211645486                     2 files


Oh, it has a readme! Always good to read it.

### Unzip the readme to the current directory

The readme file is html. Luckily, we are working in an html environment. 

### Display the contents of `readme.html` within the notebook
(Hint: check out IPython.display.IFrame)

That's some very good documentation!

### Summary: 

The files within the zip are " quoted csv's. They contain information on timeliness of departures in the US, at the departure level.

Pretty unreadable, so we go for a tool designed specifically for tabular data: **pandas**

### Load some of the data (one of the files) into memory as a pandas dataframe. What functions do you need to use?

Pro tip: there is no need to decompress the whole file. Check out zipfile.ZipFile

First, open a connection to one of the files

In [59]:
import pandas as pd
import zipfile

zip_file = zipfile.ZipFile(path_to_zips + 'On_Time_On_Time_Performance_2015_1.zip')

In [60]:
csv_file = zip_file.open('On_Time_On_Time_Performance_2015_1.csv')

Now we're ready to load the file into memory as a pandas dataframe. Remember to close the connections to the files!

In [61]:
df = pd.read_csv(csv_file)

Let's start examining the data: show the beginning of the file. How many records does it contain?

In [62]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate UniqueCarrier  \
0  2015        1      1           1          4  2015-01-01            AA   
1  2015        1      1           2          5  2015-01-02            AA   
2  2015        1      1           3          6  2015-01-03            AA   
3  2015        1      1           4          7  2015-01-04            AA   
4  2015        1      1           5          1  2015-01-05            AA   

   AirlineID Carrier TailNum      ...       Div4TailNum  Div5Airport  \
0      19805      AA  N787AA      ...               NaN          NaN   
1      19805      AA  N795AA      ...               NaN          NaN   
2      19805      AA  N788AA      ...               NaN          NaN   
3      19805      AA  N791AA      ...               NaN          NaN   
4      19805      AA  N783AA      ...               NaN          NaN   

   Div5AirportID  Div5AirportSeqID Div5WheelsOn Div5TotalGTime  \
0            NaN               NaN          NaN            NaN   
1            NaN               NaN          NaN            NaN   
2            NaN               NaN          NaN            NaN   
3            NaN               NaN          NaN            NaN   
4            NaN               NaN          NaN            NaN   

  Div5LongestGTime  Div5WheelsOff Div5TailNum  Unnamed: 109  
0              NaN            NaN         NaN           NaN  
1              NaN            NaN         NaN           NaN  
2              NaN            NaN         NaN           NaN  
3              NaN            NaN         NaN           NaN  
4              NaN            NaN         NaN           NaN  

[5 rows x 110 columns]

In [63]:
len(df)

469968

In [64]:
len(df.columns)

110

## Trimming the data

The table is quite wide, and it seems that there are many columns without much data. Which, exactly, are those? (let's consider empty a column that doesn't contain at least 1000 records, arbitrarily)

In [65]:
non_null_counts = df.count()

In [52]:
non_null_counts[non_null_counts<1000].index

Index([u'DivReachedDest', u'DivActualElapsedTime', u'DivArrDelay',
       u'DivDistance', u'Div1WheelsOff', u'Div1TailNum', u'Div2Airport',
       u'Div2AirportID', u'Div2AirportSeqID', u'Div2WheelsOn',
       u'Div2TotalGTime', u'Div2LongestGTime', u'Div2WheelsOff',
       u'Div2TailNum', u'Div3Airport', u'Div3AirportID', u'Div3AirportSeqID',
       u'Div3WheelsOn', u'Div3TotalGTime', u'Div3LongestGTime',
       u'Div3WheelsOff', u'Div3TailNum', u'Div4Airport', u'Div4AirportID',
       u'Div4AirportSeqID', u'Div4WheelsOn', u'Div4TotalGTime',
       u'Div4LongestGTime', u'Div4WheelsOff', u'Div4TailNum', u'Div5Airport',
       u'Div5AirportID', u'Div5AirportSeqID', u'Div5WheelsOn',
       u'Div5TotalGTime', u'Div5LongestGTime', u'Div5WheelsOff',
       u'Div5TailNum', u'Unnamed: 109'],
      dtype='object')

In [67]:
df.columns[df.count()<1000]

Index([u'DivReachedDest', u'DivActualElapsedTime', u'DivArrDelay',
       u'DivDistance', u'Div1WheelsOff', u'Div1TailNum', u'Div2Airport',
       u'Div2AirportID', u'Div2AirportSeqID', u'Div2WheelsOn',
       u'Div2TotalGTime', u'Div2LongestGTime', u'Div2WheelsOff',
       u'Div2TailNum', u'Div3Airport', u'Div3AirportID', u'Div3AirportSeqID',
       u'Div3WheelsOn', u'Div3TotalGTime', u'Div3LongestGTime',
       u'Div3WheelsOff', u'Div3TailNum', u'Div4Airport', u'Div4AirportID',
       u'Div4AirportSeqID', u'Div4WheelsOn', u'Div4TotalGTime',
       u'Div4LongestGTime', u'Div4WheelsOff', u'Div4TailNum', u'Div5Airport',
       u'Div5AirportID', u'Div5AirportSeqID', u'Div5WheelsOn',
       u'Div5TotalGTime', u'Div5LongestGTime', u'Div5WheelsOff',
       u'Div5TailNum', u'Unnamed: 109'],
      dtype='object')

It seems that the "diverted" fields, after the first, are often empty. No big surprise, since not that many flights must be diverted more than once in a month. Let's drop those columns, since we are not that interested in those, at least for now.

In [68]:
df.drop(df.columns[df.count()<1000],axis=1,inplace=True)

In [69]:
len(df.columns)

71

Now that we have eliminated some inconvenient data columns, let's have a look at the rest:

Let's loook at how the location data is encoded.

### Select the columns that have 'Origin' in their name

Hint: we are going to use the str attribute of Series and Indexes

In [84]:
col_contains_origin = df.columns[df.columns.str.contains('Origin')]

Let's now have a look at these. We have been using the .head() method, but that always returns the same values, so it's not that good for getting a feel for the data. Let's start using a better one, that will give us a sample of the data: df.sample()

In [71]:
df.sample(5)

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate UniqueCarrier  \
124050  2015        1      1          30          5  2015-01-30            US   
203278  2015        1      1           4          7  2015-01-04            WN   
346491  2015        1      1          19          1  2015-01-19            US   
20968   2015        1      1          26          1  2015-01-26            AA   
187647  2015        1      1          18          7  2015-01-18            WN   

        AirlineID Carrier TailNum        ...         FirstDepTime  \
124050      20355      US  N804AW        ...                  NaN   
203278      19393      WN  N215WN        ...                  NaN   
346491      20355      US  N664AW        ...                  NaN   
20968       19805      AA  N553AA        ...                  NaN   
187647      19393      WN  N746SW        ...                  NaN   

        TotalAddGTime  LongestAddGTime  DivAirportLandings Div1Airport  \
124050            NaN              NaN                   0         NaN   
203278            NaN              NaN                   0         NaN   
346491            NaN              NaN                   0         NaN   
20968             NaN              NaN                   0         NaN   
187647            NaN              NaN                   0         NaN   

       Div1AirportID Div1AirportSeqID  Div1WheelsOn Div1TotalGTime  \
124050           NaN              NaN           NaN            NaN   
203278           NaN              NaN           NaN            NaN   
346491           NaN              NaN           NaN            NaN   
20968            NaN              NaN           NaN            NaN   
187647           NaN              NaN           NaN            NaN   

        Div1LongestGTime  
124050               NaN  
203278               NaN  
346491               NaN  
20968                NaN  
187647               NaN  

[5 rows x 71 columns]

So much redundant information!!!!! Let's drop a few of those fields. Remember, 90% of the time spent in Data Science is data cleaning... 

### Do the same for the destination columns

In [85]:
col_contains_dest = df.columns[df.columns.str.contains('Dest')]

### Now, discard all these columns with redundant information. 

The set of columns we want to discard is 

`df.columns[col_contains_dest] + df.columns[col_contains_origin] - columns_of_interest`

We'll also use the opportunity to drop Year, Month, Day and Quarter columns since that is information that is already contained in FlightDate. We'll keep the DayOfWeek, though.

In [90]:
columns_of_interest = pd.Index(['Origin', 'OriginCityName', 'OriginStateName', 'Dest', 'DestCityName', 'DestStateName'])

In [97]:
cols_to_drop = pd.Index(['Year', 'Month', 'DayofMonth', 'Quarter'])

In [99]:
cols_to_drop = cols_to_drop | col_contains_dest | col_contains_origin

In [100]:
cols_to_drop = cols_to_drop.difference(columns_of_interest)

In [102]:
df.drop(cols_to_drop,axis=1,inplace=True)

OK, so we have cleaned up most of the date / location information,  and some of the extra information that was not really useful, in the process reducing the dataframe from 110 columns to 60. Let's look at the really interesting info: delays.

### Select all columns that contain 'Delay' in their column name

In [104]:
df.columns[df.columns.str.contains('Delay')]

Index([u'DepDelay', u'DepDelayMinutes', u'DepartureDelayGroups', u'ArrDelay',
       u'ArrDelayMinutes', u'ArrivalDelayGroups', u'CarrierDelay',
       u'WeatherDelay', u'NASDelay', u'SecurityDelay', u'LateAircraftDelay'],
      dtype='object')

In [106]:
df[df.columns[df.columns.str.contains('Delay')]].sample(5)

DepDelay  DepDelayMinutes  DepartureDelayGroups  ArrDelay  \
381718       NaN              NaN                   NaN       NaN   
33925          2                2                     0         6   
201728        -1                0                    -1        11   
424646        -2                0                    -1       -12   
432362         0                0                     0        -7   

        ArrDelayMinutes  ArrivalDelayGroups  CarrierDelay  WeatherDelay  \
381718              NaN                 NaN           NaN           NaN   
33925                 6                   0           NaN           NaN   
201728               11                   0           NaN           NaN   
424646                0                  -1           NaN           NaN   
432362                0                  -1           NaN           NaN   

        NASDelay  SecurityDelay  LateAircraftDelay  
381718       NaN            NaN                NaN  
33925        NaN            NaN                NaN  
201728       NaN            NaN                NaN  
424646       NaN            NaN                NaN  
432362       NaN            NaN                NaN

OK, we are going to center our attention in the DepDelay column, so we won't need the others. 

### Drop those too. 

We already know the drill, don't we?

In [111]:
cols_to_drop = df.columns[df.columns.str.contains('Delay')]
cols_to_drop = cols_to_drop.difference(['DepDelay'])
df.drop(cols_to_drop,axis=1, inplace=True)

We are not that interested in the 'Diverted' columns either, so let's drop those too

In [114]:
cols_to_drop = df.columns.str.contains('Div')
df.drop(df.columns[cols_to_drop],axis=1,inplace=True)

All of this is getting a little repetitive and boring, so let's just specify the columns we want:

In [115]:
df = df[['FlightDate', 'DayOfWeek', 'Carrier', 'TailNum', 'FlightNum', 'Origin', 
         'OriginCityName', 'OriginStateName', 'Dest', 'DestCityName', 'DestStateName',
         'DepTime', 'DepDelay', 'AirTime', 'Distance']]
df.sample(5)

FlightDate  DayOfWeek Carrier TailNum  FlightNum Origin  \
367822  2015-01-27          2      DL  N365NW       2275    AUS   
145476  2015-01-06          2      WN  N268WN       4917    LAX   
349964  2015-01-23          5      US  N809AW        483    PHX   
304168  2015-01-18          7      UA  N435UA        264    SLC   
425706  2015-01-04          7      F9  N938FR        422    DEN   

            OriginCityName OriginStateName Dest        DestCityName  \
367822          Austin, TX           Texas  SLC  Salt Lake City, UT   
145476     Los Angeles, CA      California  LAS       Las Vegas, NV   
349964         Phoenix, AZ         Arizona  GEG         Spokane, WA   
304168  Salt Lake City, UT            Utah  IAH         Houston, TX   
425706          Denver, CO        Colorado  ATL         Atlanta, GA   

       DestStateName  DepTime  DepDelay  AirTime  Distance  
367822          Utah      809        -6      148      1086  
145476        Nevada     1412        42       41       236  
349964    Washington     1749       -11      151      1020  
304168         Texas      801        -4      149      1195  
425706       Georgia        5         6      141      1199

## Formatting columns and parsing dates and times

Hurray! we have almost cleaned our dataset. Soon we will begin to do some actual work with it. 

Pandas interpreted the Deptime column as ints and the FlighDate column as strings. We want to combine them and parse them into a DateTime column, so that we can use them properly as datetimes.

First, 

### Define a function that will parse our int hours into a reasonable format ("HH:MM")

Now, we can use that function to build datetime strings that we will then pass to pd.to_datetime, with a format we will specify. Let's do that

Hint: Check out pd.to_datetime

### Overwrite the 'DepTime' column with its version in the proper format

Hint: Before overwriting your column in the dataframe, make sure that everything works by assigning the modified column to a Series variable.

### Now, create a DepDateTime with the proper type using `pd.to_datetime`

### Inspect the types, see if everything is in order

### Transform the FlightNum column into a column of strings

## Examining the data

Finally clean! Let's start to do some preliminary work on the data.

### Find the biggest delays. 

How would you find the 5 maximum delays?

### What was the average delay for this month? Standard deviation and typical value?

Also: a quick look at the correlation between the numerical variables is extremely easy with pandas:

### A little taste of groupby

Very often, we will want to split our data according to a variable, then compute some statistics on the different groups. We will see this in depth next week, but I want to give you a little taste.

What is the plane that has the highest average delay? We'll first group by tail number (the *license plate* of a plane) and then calculate the relevant statistic for each group (group of *departures*)

Oh, but those numbers smell like these planes had only a few, very delayed, departures! how can we count the number of departures *and* calculate the average delay at the same time?

Awesome! now, let's look at the average delay of the planes with some departures (let's say, at least 15)

## Plotting

There are several ways in which we could go about this. 

For example, do the delays have a relationship with the number of departures a plane does?

 Another question: how do the delays stack over the course of the day?

## Self study:

Now, you can try the above with the whole dataset. First, we need to join all the zipped csvs into one dataframe. How would you do that? Do you think you could run into memory problems? How would you avoid that?

Hints for reading large files: 
* selecting columns:
    `pd.read_csv(usecols=[list_of_interesting_column_names])`
* bcolz
* specify column types